Added Spektrum SRXL2 serial protocol for new upcoming receiver support

SpektrumRC opened the pull request:
Adding new protocol for (SRXL2) for new Spektrum receivers, currently supported by the SPM4650.
This protocol requires half-duplex communication as it only uses one wire, and involves a handshaking process in order to operate, so the only way for this to be wired is a UART Tx pin.

Since there isn't enough memory to support this on F3's, I did the following in the "common_pre.h" file. I'm not sure if there is a better way to handle this.
`#ifndef STM32F3 
#define USE_SERIALRX_SRXLv2     // Spektrum SRXL2 protocol
#endif`

For reference, the initial pull request for this protocol was #8598, but it was requested that the UART changes get moved to a separate pull request.

etracer65 commented on src/main/target/common_pre.h:
```diff
@@ -175,6 +175,10 @@
 #define USE_SERIALRX_SUMD       // Graupner Hott protocol
 
+#ifndef STM32F3 //Not enough space for F3 support
+#define USE_SERIALRX_SRXLv2     // Spektrum SRXL2 protocol
```
This is non-standard. The `USE_` define should be all uppercase.

etracer65 replied:
Also the define should be added to the `#if (FLASH_SIZE > 256)` section rather than adding an additional F3 exclusion here.

SpektrumRC replied:
> This is non-standard. The `USE_` define should be all uppercase.

Understood, I've renamed all SRXLv2 references in any case to be SRXL2 ('v' removed). I think this more sense than having a capital V included with the protocol name itself, and it matches how it's named in our public documentation anyway.

SpektrumRC replied:
> Also the define should be added to the `#if (FLASH_SIZE > 256)` section rather than adding an additional F3 exclusion here.

Would it make sense to have a new `#if (FLASH_SIZE > 256)` section next to the other serial protocols in case any other protocols come down the line (would be nice to have the USE_SERIALRX_xxxx in the same section), or would that just cause more confusion down the line? I will include it at the very end of the current section otherwise.

etracer65 replied:
From a maintainability perspective I would prefer just adding it to the existing `#if (FLASH_SIZE > 256)` section and not create an additional area for maintenance. But let's hear some other opinions before changing.

SpektrumRC replied:
> From a maintainability perspective I would prefer just adding it to the existing `#if (FLASH_SIZE > 256)` section and not create an additional area for maintenance. But let's hear some other opinions before changing.

I've moved it to the larger > 256 section. I think down the line if the serial protocols want to be grouped together in a new section, it can be changed then.

SpektrumRC replied:
@etracer65 Can you mark this as resolved? I think this is the last change request preventing the merge. 
Thanks

ledvinap commented on src/main/rx/srxl2_types.h:
```diff
@@ -0,0 +1,138 @@
+    SendHandshake,
+    ListenForHandshake,
+    Running
+} srxl2State;
```
Types start with CapitalLetter in BF

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+            // @todo if there were non-handshake packets - go to running,
+            // if there were - go to either Send Handshake or Listen For Handshake
+            state = Running;
+        } else if (last_idle_timestamp > last_receive_timestamp) {
```
Would this be the suggested correction? 
`} else if (cmpTimeUs(last_idle_timestamp, lastReceiveTimestamp) > 0) {`

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+#define SRXL2_LISTEN_FOR_HANDSHAKE_TIMEOUT 200000
+
+
+static uint8_t unit_id = 0;
```
Fixed in last commit

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+
+#define SRXL2_FRAME_TIMEOUT_US         50000
+
+#define SRXL2_LISTEN_FOR_ACTIVITY_TIMEOUT 50000
```
Added _US suffix. I'm a bit hesitant to remove _TIMEOUT however. I guess I could remove that and add a comment next to the define indicating it is a timeout value?

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+#ifdef USE_SERIALRX_SRXL2
+
+#define SRXL2_MAX_CHANNELS             32
+#define SRXL2_TIME_BETWEEN_FRAMES_US   11000 // 5500 for DSMR
```
Corrected in last  commit

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+#define DEBUG(...)
+#endif
+
+#define USE_SERIALRX_SRXL2
```
Good catch. Removed in last commit

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+static volatile uint32_t last_receive_timestamp = 0;
+static volatile uint32_t last_valid_packet_timestamp = 0;
+
+static volatile uint8_t read_buffer[2 * SRXL2_MAX_PACKET_LENGTH];
```
@DieHertz Do you think these volatiles can be removed, or were they required when you worked on this?

SpektrumRC replied:
Removed voltile for variables not accessed in interrupts

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+
+    const uint16_t *frame_channels = (const uint16_t *) (channel_data + 1);
+
+    if (channel_data->channel_mask.u8.channels_0_7) {
```
Opted for outer loop for now

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+
+    //    cliPrintf("> %d %d %x\r\n", read_buffer_idx, read_buffer_read_idx, anonymous.header.id);
+
+    const uint8_t bytes_to_read = anonymous.header.length;
```
> (unsigned)

I'm sorry if this suggestion is not immediately obvious to me. You are suggesting unsigned but the declaration is already uint8_t, so what is the suggestion here exactly?

mikeller replied:
Just `unsigned` (i.e. the MCU's word size). Limiting to a smaller type will result in the compiler adding in overflow checks, and less efficient code.

SpektrumRC replied:
Ah ok, I understand now. This has been removed with the double buffer changes, but maybe the index variables should be changed to unsigned for the same reason

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+
+    //    cliPrintf("- %d %d %x\r\n", read_buffer_idx, read_buffer_read_idx, anonymous.header.id);
+
+    if (anonymous.header.id != SRXL2_ID ||
```
Removed the id check, but left the check it was or'd with .

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+
+        case FailsafeChannelData: {
+            srxl2ProcessChannelData((const srxl2ChannelDataHeader *) (control_data + 1), rxRuntimeConfig);
+            setRssiDirect(0, RSSI_SOURCE_RX_PROTOCOL);
```
So this only happens for the failsafe case. The receiver does send a rssi value of 0 with failsafe data, but it is being set by the setRssi function which goes through filtering from what I can tell. Does it make sense to make the failsafe rssi value go through the filter instead of just directly setting the value?

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+        return;
+    }
+
+    const uint16_t calculated_crc = crc16_ccitt_update(0, anonymous.local_buffer, anonymous.header.length - 2);
```
The crc doesn't work in this case if I dont truncate the length.

ledvinap replied:
That's strange .. when you calculate CRC over whole message, result should be 0 if packet has correct CRC
https://onlinegdb.com/HJg1K67uMB

SpektrumRC replied:
> That's strange .. when you calculate CRC over whole message, result should be 0 if packet has correct CRC
> https://onlinegdb.com/HJg1K67uMB

Oh sorry, I missunderstood what you were trying to do initially. It does appear to work as you stated. I will make the change

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+        return true;
+    }
+
+    if (((handshake->destination_device_id >> 4) & 0xF) != FlightController ||
```
Done. Had to add additional parentheses around `((FlightController << 4) | unit_id)` as `!=` has higher order of operation priority than bitwise operations.

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+    union {
+        uint8_t local_buffer[SRXL2_MAX_PACKET_LENGTH];
+        srxl2Header header;
+    } anonymous;
```
changed to srxl2Frame

ledvinap commented:
@SpektrumRC : Can you please avoid force-pushing until this PR is ready? Github does detach conversations, making review more complicated ...

SpektrumRC commented:
> @SpektrumRC : Can you please avoid force-pushing until this PR is ready? Github does detach conversations, making review more complicated ...

Yes, sorry for making things more difficult. I appreciate the time you're taking to review this as it is.

mikeller commented on src/main/cli/cli.c:
```diff
@@ -6090,6 +6111,9 @@ const clicmd_t cmdTable[] = {
         "\treverse <servo> <source> r|n", cliServoMix),
+#endif
+#if defined (USE_SERIALRX_SRXL2)
+    CLI_COMMAND_DEF("srxl2_bind", "SRXL2 bind", "[enter|info]", cliSrxl2Bind),
```
I think it's a more user friendly approach if we integrate this into `bind_rx_spi` above, and rename the command to `bind_rx` - after all the user will only ever have one RX configured, and should not have to care what RX type it is that they want to bind.
(Thinking ahead, we might want to add this to MSP as well, and give users a 'Bind RX' button in configurator if their RX supports it.)

SpektrumRC replied:
Cli changes for this done in the last commit. Not sure if that was appropriate, or if it should've been a separate pull request. I can undo the changes if so.
Would love to see it in MSP + button in configurator as well.

mikeller commented on src/main/cli/settings.c:
```diff
@@ -703,6 +714,10 @@ const clivalue_t valueTable[] = {
+#endif
+#ifdef USE_SERIALRX_SRXL2
+    { "srxl2_unit_id",             VAR_UINT8  | MASTER_VALUE, .config.minmax = { 0, 0xf }, PG_RX_CONFIG, offsetof(rxConfig_t, srxl2_unit_id) },
+    { "srxl2_baud_rate",           VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_SRXL2_BAUD_RATES }, PG_RX_CONFIG, offsetof(rxConfig_t, srxl2_baud_rate) },
```
Is there a plan to add support for bit rates different to 115k and 400k? If not, then this should be done as a `TABLE_OFF_ON` setting `srxl2_speed_115k` or `srxl2_speed_slow` or suchlike.
(This is only a temporary solution anyway, as this will be amalgamated into the Unified UART configuration once this is being done.)

SpektrumRC replied:
Currently we dont have "plans" to. The protocol was developed to support the addition of new bit rates though, should there be a need to have faster speed down the line. These new bit rates would be pre-defined however, rather than being defined as an integer, so I dont mind making it a table that would be srxl2_speed_115k, srxl2_speed_400k, etc.

SpektrumRC replied:
Sorry, I just looked at this and realized it was already in a table. Were you implying to just change the "values" from 115k  to srxl2_speed_115k style?

mikeller replied:
If it's only 2 baud rates for now, it should be using the `TABLE_OFF_ON` enum. As I said, this is only temporary, once we redo the UART configuration to allow setting UART parameters by UART instead of by protocol using it, setting the bit rate for SRXL2 will be done by setting the bit rate for the UART it is using.

SpektrumRC replied:
I can change it to TABLE_OFF_ON since it will be temporary anyway. When the change is made to set bit rate by uart rather than protocol, will those bitrate options be an enum with the selection options limited by the protocol, or how will that be handled so only supported bit rates are selectable? Will it also be possible to have default bit rates selected based on protocol selected?

mikeller replied:
Thanks for changing.
I think we will have to look into how we can support protocols for which the available hardware only supports certain speeds in a way that makes it easy for the user to select the correct speed - Spektrum protocols aren't the only ones that are affected by this.
Potentially this will be done by allowing setting of all possible speeds in the firmware / CLI, but restricting the speeds that are selectable in configurator, based on the protocol that has been selected for the UART. This also caters to situations where the 'official' hardware only supports certain speeds, but open source / aftermarket implementations of the same protocol support different speeds.

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+
+#if SRXL2_DEBUG
+void cliPrintf(const char *format, ...);
+#define DEBUG(format, ...) cliPrintf(format, __VA_ARGS__)
```
Defining a dependency on `cli.c` here is risky and likely to break when the CLI code is refactored.

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+        setRssi(scaleRange(rssiPercent, 0, 100, 0, RSSI_MAX_VALUE), RSSI_SOURCE_RX_PROTOCOL);
+    } else {
+        const int rssiDbm = channelData->rssi;
+        (void) rssiDbm;
```
What is this for?

SpektrumRC replied:
The protocol can provide either a positive percent value for rssi (receiver handles converting dBm to %), or a negative dBm value. This else statement appears to have been a placeholder for how to handle the negative dBm value. Not sure if dBm is something betaflight tends to handle as it is usually displayed in %.  I would add in a default conversion table, but not all receivers have the same sensitivity range, so I think I'd prefer to remove the first 2 lines, and just have it set the rssi to 50% as it currently does.

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+        }
+    }
+
+    // DEBUG("channel data: %d %d %x\r\n",
```
Please don't leave commented out code in without an annotation - state when and under what circumstances it should be re-enabled, or remove it.

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+bool srxl2ProcessPacket(const Srxl2Header* header, rxRuntimeConfig_t *rxRuntimeConfig)
+{
+    switch (header->packetType) {
+    case Handshake: return srxl2ProcessHandshake(header, (const Srxl2HandshakeSubHeader *) (header + 1));
```
Can we have these on separate lines for consistency with the rest of the code please?

SpektrumRC replied:
Done

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+        }
+    }
+
+    // @todo find out why it happens
```
Is this still open?

SpektrumRC replied:
Closed now

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+        union {
+            uint8_t localBuffer[sizeof(Srxl2Header)];
+            Srxl2Header header;
+        } srxl2Frame;
```
This union is defined locally in a number of functions - having one unit wide definition will make this cleaner.

SpektrumRC replied:
Double buffer changes have defined a modified version of this union globally, and used it to declare the buffers. Please verify if this is acceptable.

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+        }
+    }
+
+    /* handshake protocol
```
In terms of maintainability I think it is best if helpful comments such as this one are moved to the top of the file, where other developers will see them before they start reading code.

SpektrumRC replied:
Done.

mikeller commented on src/main/target/common_pre.h:
```diff
@@ -175,6 +175,7 @@
 #define USE_SERIALRX_SPEKTRUM   // SRXL, DSM2 and DSMX protocol
 #define USE_SERIALRX_SUMD       // Graupner Hott protocol
 
+
```
:space_invader:

mikeller commented on src/main/telemetry/srxl.c:
```diff
@@ -83,24 +84,33 @@
-    // write the telemetry frame to the receiver.
-    srxlRxWriteTelemetryData(sbufPtr(dst), sbufBytesRemaining(dst));
+    if (srxl2) {
+      srxl2FinalizeFrame(dst);
```
:space_invader:

(Indentation in existing code in this file seems to be dodgy. I think it's best if we leave it as it is for this pull request to facilitate reviewing, and then open a new pull request with indentation fixes only once this has been merged.)

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+
+    // @todo find out why it happens
+    if (anonymous.header.id != SRXL2_ID) {
+      read_buffer_read_idx = read_buffer_idx;
```
See double buffer changes

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+
+    uint32_t bytes_copied = 0;
+    while (bytes_copied < sizeof(srxl2Header)) {
+        anonymous.local_buffer[bytes_copied++] = read_buffer[read_buffer_read_idx++];
```
Changed to double buffer

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+    last_receive_timestamp = microsISR();
+
+    // circular buffer depleted and waiting for sync character
+    if (read_buffer_read_idx == read_buffer_idx && character != SRXL2_ID) {
```
Changed to double buffer

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+            ? read_buffer_idx + (uint8_t) sizeof(read_buffer) - read_buffer_read_idx
+            : read_buffer_idx - read_buffer_read_idx;
+
+    if (bytes_available >= sizeof(srxl2Header)) {
```
Header no longer read until a packet is ready as indicated by idle interrupt.

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,634 @@
+            return RX_FRAME_DROPPED;
+        }
+
+        if (bytes_available < anonymous.header.length) {
```
Header no longer read until a packet is ready as indicated by idle interrupt.

SpektrumRC commented on src/main/cli/cli.c:
```diff
@@ -357,6 +358,8 @@ static void cliPutp(void *p, char ch)
 
 static void cliPrintfva(const char *format, va_list va)
 {
+    if (!cliWriter) return;
```
Noted, will remove the cli changes related to debugging once all other fixes are approved.

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+#define SRXL2_PORT_OPTIONS             (SERIAL_STOPBITS_1 | SERIAL_PARITY_NO | SERIAL_BIDIR)
+#define SRXL2_PORT_MODE                MODE_RXTX
+
+#define SRXL2_REPLY_QUIESCENCE         (2 * 10 * 1000000 / SRXL2_PORT_BAUDRATE_DEFAULT) /*2 * (lastIdleTimestamp - lastReceiveTimestamp)*/
```
Done

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+static uint8_t baudRate = 0;
+
+static Srxl2State state = Disabled;
+static volatile uint32_t timeoutTimestamp = 0;
```
Have removed volatile declaration for variables not accessed within interrupts.

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+static uint8_t telemetryFrame[22];
+
+uint8_t globalResult = 0;
+bool srxl2ProcessHandshake(const Srxl2Header* header, const Srxl2HandshakeSubHeader* handshake)
```
Done

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+
+    const uint16_t *frameChannels = (const uint16_t *) (channelData + 1);
+
+    for(size_t i = 0; i<4; i++)
```
Done

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+
+    //    cliPrintf("- %d %d %x\r\n", readBufferIdx, readBufferReadIdx, srxl2Frame.header.id);
+
+    if (bytesCopied != srxl2Frame.header.length) {
```
Removed as part of double buffer changes

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+        return;
+    }
+
+    const uint16_t calculatedCrc = crc16_ccitt_update(0, srxl2Frame.localBuffer, srxl2Frame.header.length - 2);
```
Done

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+
+    if (lastIdleTimestamp > lastReceiveTimestamp) {
+        // time sufficient for at least 2 characters has passed
+        if (now - lastReceiveTimestamp > SRXL2_REPLY_QUIESCENCE) {
```
I've done some tests with telemetry getting handled sooner than it currently does, and this part is still required to guarantee the idle period required. With the speed at which telemetry is currently handled however, it doesn't seem to be necessary, but I'd like to keep this in case that changes

ledvinap replied:
You can set idle timeout delay to value equal or slightly smaller than telemetry delay. Then it is always possible to transmit just after timeout ISR

SpektrumRC replied:
Could you elaborate more on how to do this? It sounds like you're saying there is a way to configure the idle timeout for the uart peripheral, but I cant find any registers to do so. Based on my interpretation of the STM32 documentation, the idle interrupt delay is fixed to one byte worth of time where the signal line is idle (high for non inverted uart). This means the delay will also vary based on baud rate.

ledvinap replied:
I opened F3 documentation I have locally, USARTx_RTOR is available. It is possible cores on different CPUs don't support this (and IMO F1 doesn't support it at all)

SpektrumRC replied:
I was referencing the reference manual for the STM32F405 and I cant find any references to that register. It seems it is only supported on some MCU's then.  With that given, would this existing implementation be acceptable, or do you have another suggestion on how this could be addressed better?

ledvinap replied:
Yes, peripheral codes are different. F3 is analog-oriented, so it's cores are usually more advanced.

There isn't much to improve now ... 
just write is as `if (cmpTimeUs(now, lastReceiveTimestamp) > SRXL2_REPLY_QUIESCENCE)` 
or `if (lastReceiveTimestamp && cmpTimeUs(now, lastReceiveTimestamp) > SRXL2_REPLY_QUIESCENCE)` if lastReceiveTimestamp may be zeroed until first packet is received (but first version will fail only if no packet was received for about 30 minutes after power up)

`cmpTimeUs` is just wrapper around unsigned subtraction converted to signed,  but it's obvious that timestamps are compared. Resulting value is signed difference between timestamps, with range of about +-30 minutes (2^31 uS) and correct timestamp wrap handling 
(simple subtraction is not portable; cmp implementation is also not portable, but works on ARM (and most modern processors) and can be easily fixed in single place is necessary)

SpektrumRC replied:
your first  suggestion should work fine. `lastReceiveTimestamp` is set per byte reception so packetes dont have to be valid, and assuming a receiver is connected physically, it will always be sending data regardless of being bound or not.

SpektrumRC replied:
Changed in commit 4

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,610 @@
+
+    const uint32_t now = micros();
+
+    if (lastIdleTimestamp > lastReceiveTimestamp) {
```
done

ledvinap commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+    serialPort = openSerialPort(portConfig->identifier, FUNCTION_RX_SERIAL, srxl2DataReceive,
+        NULL, SRXL2_PORT_BAUDRATE_DEFAULT, SRXL2_PORT_MODE, options);
+
+    if (serialPort) {
```
`if (!serialPort) return false; ` is consistent with code above - return on error / continue processing  without nesting

SpektrumRC replied:
done in commit 4

ledvinap commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+        }
+    }
+
+    return serialPort;
```
(implicit conversion to bool works, but is a bit confusing)

SpektrumRC replied:
cast to bool in commit 4

ledvinap commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+    };
+
+    const uint16_t crc = crc16_ccitt_update(0, &bind, length - 2);
+    bind.crcHigh = ((uint8_t *) &crc)[1];
```
>>8 , &0xff

ledvinap commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+        }
+    };
+
+    const uint16_t crc = crc16_ccitt_update(0, &bind, length - 2);
```
CRC calculation is repeated 3 times in this PR, it may be implemented in srxl2RxWriteData (I assume all frames have CRC)

SpektrumRC replied:
Good idea. Crc calculation for all SRXL2 writes are handled in that function in commit 4

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+        const int rssiPercent = channelData->rssi;
+        setRssi(scaleRange(rssiPercent, 0, 100, 0, RSSI_MAX_VALUE), RSSI_SOURCE_RX_PROTOCOL);
+    } else {
+        // If dBm value provided, cant properly convert to % without knowing the receivers sensitivity range. Fix at 50% for now.
```
My concern here is that having a built in conversion, even if not the same as our provided conversion, might not be enough to make users feel like something is wrong. If they have the option to change the receivers output to dBm instead of %, then they should be able to set it back. By default, all new receivers should be providing %, and we may give them the option to change this somehow down the line. If rssi doesn't work at all, that is a clear indicator to them that something is not set up correctly, and something has to be changed. I prefer this over inconsistent behavior without them fully realizing it. It would be nicer if betaflight could show dBm instead of % is that is what the receiver is providing.

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+
+    const uint16_t *frameChannels = (const uint16_t *) (channelData + 1);
+
+    for (size_t i = 0; i < 4; i++) {
```
I like that. Changed to this style in commit 4.

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+    }
+
+    switch (controlData->command) {
+    case ChannelData: {
```
Fixed in commit 4

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+    case Handshake: 
+        return srxl2ProcessHandshake(header);
+    case ControlData: 
+        return srxl2ProcessControlData((const Srxl2ControlDataSubHeader *) (header + 1), rxRuntimeConfig);
```
Makes sense. Changed to this in commit 4.

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+
+    readBufferIdx = 0;
+    //Clear readBuffer for next read
+    memset((uint8_t*)readBufferPtr, 0, SRXL2_MAX_PACKET_LENGTH);
```
Removed in commit 4

SpektrumRC commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,558 @@
+
+    case Running: {
+        // frame timed out, reset state
+        if (lastValidPacketTimestamp < now && now - lastValidPacketTimestamp >= SRXL2_FRAME_TIMEOUT_US) {
```
Replaced a bunch of these with cmpTimeUS in commit 4

SpektrumRC commented:
> Much better now, thanks for your patience.
> The double-buffering code is quite prone to race conditions, but that should be quite easy to fix, I'll look into it later. It should be good enough for testing.

No, thank you guys for your patience looking through this and providing feedback/necessary changes to get this merged. I really appreciate it.

I've made 2 more commits. I believe they address most if not all standing issues now. Please let me know if anything else remains.

ledvinap commented:
One possible way to improve receive consistency (probably not best one, but it's quite simple). It's only rough sketch (it would be probably easier to write actual code)

Receive buffer:
```
struct rxBuf {
  volatile uint8_t len;
  Srxl2Frame packet;
}
struct rxBuf readBuffer[2]; // possibly use more buffers
```
`rxBuff.len == 0` : Buffer is available/used in ISR, receive in progress
`rxBuff.len >=0` : Packet is ready/processing in main task

Serial receive callback:
Store data using (ISR owned) buffer pointer, checking length. If pointer is NULL, do nothing. If length overflow, disable reception (and drop packet).

Serial idle callback:
If there are received data(and reception not disabled), write rxBuf.len (signaling main task that packet is complete), switch to next buffer.
If (possibly new buffer) `rxBuf.len > 0` - disable reception, ignore all chars until next idle callback (and do not switch buffers on next idle)
Else setup pointer for receive ISR / enable reception

It is possible to postpone`rxBuf.len == 0` test until first character is received - it will give more time to main task to handle packet, possibly working with single buffer. But it will complicate code and with reasonable processing speed both should be equivalent 

Packet processing:
Wait until current `rxBuff.len > 0`
process packet
set `rxBuff.len = 0;` - receive buffer is available for ISR
switch to next buffer

mikeller commented on src/main/cli/cli.c:
```diff
@@ -3246,37 +3247,52 @@ static void cliBeeper(char *cmdline)
+            break;
+#endif
+        }
+    } else if (featureIsEnabled(FEATURE_RX_SPI)) {
```
This `else if` should be guarded by a `#if defined(USE_RX_SPI)` now.

SpektrumRC replied:
Done

mikeller commented on src/main/cli/settings.c:
```diff
@@ -703,6 +704,10 @@ const clivalue_t valueTable[] = {
+#endif
+#ifdef USE_SERIALRX_SRXL2
+    { "srxl2_unit_id",             VAR_UINT8  | MASTER_VALUE, .config.minmax = { 0, 0xf }, PG_RX_CONFIG, offsetof(rxConfig_t, srxl2_unit_id) },
+    { "srxl2_baud_rate",           VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_OFF_ON }, PG_RX_CONFIG, offsetof(rxConfig_t, srxl2_baud_rate) },
```
I'd rename this to `srxl2_baud_fast` or similar - from the user's point of view, `set srxl2_baud_rate = on` is rather cryptic.

SpektrumRC replied:
Done

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,545 @@
+#include "platform.h"
```
This file should have the same header as https://github.com/betaflight/betaflight/pull/8606/files#diff-2da41031eb7d7f076567ebda42e5223e

SpektrumRC replied:
Done

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,545 @@
+#include "drivers/serial.h"
+#include "drivers/serial_uart.h"
+
+#include <string.h>
```
System libraries go to the top, before `platform.h`.

SpektrumRC replied:
Done

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,545 @@
+#include "rx/srxl2_types.h"
+#include "io/spektrum_vtx_control.h"
+
+#include "drivers/serial.h"
```
Should be together with the rest of `driver/`.

SpektrumRC replied:
Done

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,545 @@
+#define DEBUG(...)
+#endif
+
+#ifdef USE_SERIALRX_SRXL2
```
This should go right underneath `platform.h` (which includes the target specific defines).

SpektrumRC replied:
Done

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,545 @@
+    readBufferIdx = 0;
+}
+
+uint32_t autobaud_timeout = 0;
```
This seems to be unused now.

SpektrumRC replied:
Removed

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -0,0 +1,545 @@
+        return 0;
+    }
+
+    return 988 + ((rxRuntimeConfig->channelData[channelIdx] >> SRXL2_CHANNEL_SHIFT) >> 1);
```
Using a define here instead of a magic number will make the code easier to read / maintain.

SpektrumRC replied:
Done

SpektrumRC commented:
Just made another commit that addresses the issues you guys brought up. Mostly formatting. Largest change was the rxBuf struct usage as described by @ledvinap. I implemented it as best as I could interpret your suggestions, but I am primarily a hardware guy after all, so please let me know if there is anything I did that doesn't make sense.
Also, in addition to checking the buffer length to determine if it's ready to be processed, I added a NULL ptr check first. Not sure if it's really necessary, dont think there will every be a situation in which the ptr is NULL.

If this is good, I can combine all the commits into one and force push before the pull request is merged, unless that's not necessary?

SpektrumRC commented:
Also, I just noticed I failed a build test for SPRACINGH7EXTREME. I dont recall this being a problem before? It says the RAM overflowed by 288 bytes. I cant think of any of my recent changes having taken up that much more ram than before. Is this a new target that just wasn't there before when I made this pull request initially?

mikeller commented:
@SpektrumRC: Looking good now. Can you please rebase / squash into one commit? I've opened #8668 to address the RAM overflow on the SPRACINGH7EXTREME, and the checks on this should come right once they are run after #8668 has been merged.

SpektrumRC commented:
> @SpektrumRC: Looking good now. Can you please rebase / squash into one commit? I've opened #8668 to address the RAM overflow on the SPRACINGH7EXTREME, and the checks on this should come right once they are run after #8668 has been merged.

Rebase and squash done. Thanks again for taking the time taken to review and request necessary changes!

etracer65 commented on src/main/target/common_pre.h:
```diff
@@ -333,4 +333,5 @@
 #define USE_VTX_TABLE
 #define USE_PERSISTENT_STATS
 #define USE_PROFILE_NAMES
+#define USE_SERIALRX_SRXL2     // Spektrum SRXL2 protocol
```
There's a problem with code missing proper bounding with `USE_SERIALRX_SRXL2`.

Disabling the `#define USE_SERIALRX_SRXL2` will result in compilation errors. Likely you have some unbounded references to code that is only included when `USE_SERIALRX_SRXL2` is defined.

```
/var/folders/75/1m9wjhn08xj5g10059b9s2sr0000gn/T//ccquXECb.ltrans2.ltrans.o: In function `taskTelemetry.lto_priv.420':
<artificial>:(.text+0x469e): undefined reference to `srxl2TelemetryRequested'
/var/folders/75/1m9wjhn08xj5g10059b9s2sr0000gn/T//ccquXECb.ltrans6.ltrans.o: In function `processSrxl.lto_priv.1131':
<artificial>:(.text+0x5ae2): undefined reference to `srxl2InitializeFrame'
<artificial>:(.text+0x5b5a): undefined reference to `srxl2FinalizeFrame'
/var/folders/75/1m9wjhn08xj5g10059b9s2sr0000gn/T//ccquXECb.ltrans6.ltrans.o: In function `telemetryInit':
<artificial>:(.text+0x64b2): undefined reference to `srxl2RxIsActive'
```

SpektrumRC replied:
@etracer65 Is it ok if I do a force push for this fix or would it be better for the initial commit to be separate, and then squash?

SpektrumRC replied:
Went ahead and just made a separate commit for now. Let me know if I can squash

etracer65 replied:
Looks good. You can go ahead and squash.

SpektrumRC replied:
Done and pushed.
